tracing/tracing.go

```diff
@@ -0,0 +1,149 @@
+	// it will be using the parent sampler or the default.
+	sampler = trace.NeverSample()
+)
+
+func init() {
+	enableFromEnv()
+}
+
+func enableFromEnv() {
+	_, ok := os.LookupEnv("AZURE_SDK_TRACING_ENABELD")
+	if ok {
+		agentEndpoint, ok := os.LookupEnv("OCAGENT_TRACE_EXPORTER_ENDPOINT")
+
+		if ok {
+			EnableWithAIForwarding(agentEndpoint)
+		} else {
+			Enable()
+		}
+	}
+}
+
+// Enable will start instrumentation for metrics and traces.
+func Enable() (err error) {
+	Enabled = true
+	sampler = nil
+	Transport = &ochttp.Transport{Propagation: &tracecontext.HTTPFormat{}}
+
+	if err != nil {
+		return
+	}
+	err = initStats()
+	return
+}
+
+// EnableWithAIForwarding will start instrumentation and will connect to app insights forwarder
+// exporter making the metrics and traces available in app insights.
+func EnableWithAIForwarding(agentEndpoint string) (err error) {
+	err = Enable()
+	if err != nil {
+		return err
+	}
+
+	exporter, err := ocagent.NewExporter(ocagent.WithInsecure(), ocagent.WithAddress(agentEndpoint))
+	if err != nil {
+		return err
+	}
+	trace.RegisterExporter(exporter)
+	return
+}
+
+// initStats registers the views for the http metrics
+func initStats() (err error) {
+	clientViews := []*view.View{
+		ochttp.ClientCompletedCount,
+		ochttp.ClientRoundtripLatencyDistribution,
+		ochttp.ClientReceivedBytesDistribution,
+		ochttp.ClientSentBytesDistribution,
+	}
+	if err = view.Register(clientViews...); err != nil {
+		return err
+	}
+	return
+}
+
+// StartSpan starts a trace span
+func StartSpan(ctx context.Context, name string) context.Context {
+	ctx, _ = trace.StartSpan(ctx, name, trace.WithSampler(sampler))
+	return ctx
+}
+
+// EndSpan ends a previously started span stored in the context
+func EndSpan(ctx context.Context, httpStatusCode int, err error) {
+	span := trace.FromContext(ctx)
+
+	if span == nil {
+		return
+	}
+
+	if err != nil {
+		span.SetStatus(trace.Status{Message: err.Error(), Code: toTraceStatusCode(httpStatusCode)})
+	}
+	span.End()
+}
+
+// TraceStatusCode converts HTTP Codes to OpenCensus codes as defined
+// at https://github.com/census-instrumentation/opencensus-specs/blob/master/trace/HTTP.md#status
+func toTraceStatusCode(httpStatusCode int) (code int32) {
+	switch httpStatusCode {
+	case 499:
+		return trace.StatusCodeCancelled
+	case http.StatusBadRequest:
+		return trace.StatusCodeInvalidArgument
+	case http.StatusGatewayTimeout:
+		return trace.StatusCodeDeadlineExceeded
+	case http.StatusNotFound:
+		return trace.StatusCodeNotFound
+	case http.StatusForbidden:
+		return trace.StatusCodePermissionDenied
+	case http.StatusUnauthorized: // 401 is actually unauthenticated.
+		return trace.StatusCodeUnauthenticated
+	case http.StatusTooManyRequests:
+		return trace.StatusCodeResourceExhausted
+	case http.StatusNotImplemented:
+		return trace.StatusCodeUnimplemented
+	case http.StatusServiceUnavailable:
+		return trace.StatusCodeUnavailable
+	case http.StatusOK:
```

Comment:
Is this correct?  According to the table (linked in the func's comment) 200-399 should be considered OK.

Reply:
yes, its a shame the library doesnt provide this functionality, ill make a PR upstream to add it after this